site-src/geps/gep-1016.md

```diff
@@ -0,0 +1,813 @@
+- If service and method are supplied and the type is `Exact`, translate to a
+  path match of `/SERVICE/METHOD` with type `Exact`
+- If service is supplied but not method and the type is `Exact`, translate to a
+  path match of `/SERVICE/` with type `PathPrefix`
+- Otherwise, translate the service and the method individually into regular
+  expressions (with an empty field interpreted as `.*`), and translate to an
+  HTTPRoute path matcher of type `RegularExpression` with value `/SERVICE_REGEX/MATCHER_REGEX`
+
+
+#### An Example Implementation
+
+In the interest of proving the feasibility of implementation of this proposal,
+we include a sketch of an implementation for an xDS-based control plane.
+
+Suppose the following resources are applied to the cluster:
+
+```yaml
+apiVersion: gateway.networking.k8s.io/v1alpha2
+kind: Gateway
+metadata:
+  name: my-gateway
+spec:
+  gatewayClassName: acme-lb
+  listeners:
+  - name: http
+    protocol: HTTP
+    port: 80
+---
+kind: GRPCRoute
+apiVersion: gateway.networking.k8s.io/v1alpha2
+metadata:
+  name: foo-grpcroute
+spec:
+  parentRefs:
+  - name: my-gateway
+  hostnames:
+  - foo.com
+  rules:
+    matches:
+    - method:
+        service: foo.bar
+    backendRefs:
+    - name: grpc-backend
+      port: 50051
+---
+kind: HTTPRoute
+apiVersion: gateway.networking.k8s.io/v1alpha2
+metadata:
+  name: foo-httproute
+spec:
+  parentRefs:
+  - name: my-gateway
+  hostnames:
+  - foo.com
+  rules:
+    matches:
+    - path:
+        type: PathPrefix
+        value: /foo
+    backendRefs:
+    - name: http-backend
+      port: 80
+```
+
+According to the above scheme, we translate the gRPC method matcher to an HTTP
+path matcher of type `PathPrefix` and value `/foo.bar/`. Following the HTTPRoute
+precedence rules, `/foo.bar/` has more characters than `/foo`, so it has higher
+precedence. Hence, the GRPCRoute's rule must be evaluated before the HTTPRoute's
+rule.
+
+Accordingly, an xDS server should create something like the following
+[`RouteConfiguration`](https://github.com/envoyproxy/envoy/blob/379a7567b0d15cc9d0806052b82670c8cad76fc5/api/envoy/config/route/v3/route.proto#L26):
+
+```json
+{
+  "virtual_hosts": [
+    "domains": ["foo.com"],
+    "routes": [
+      {
+        "prefix": "/foo.bar/",
+        "route": {
+          "cluster": "grpc-backend-cluster"
+        }
+      },
+      {
+        "prefix": "/foo",
+        "route": {
+          "cluster": "http-backend-cluster"
+        }
+      }
+    ]
+  ]
+}
+```
+
+## API
+
+The API deviates from `HTTPRoute` where it results in a better UX for gRPC
+users, while mirroring it in all other cases.
+
+### Example `GRPCRoute`
+
+```yaml
+kind: GRPCRoute
+apiVersion: gateway.networking.k8s.io/v1alpha2
+metadata:
+  name: foo-grpcroute
+spec:
+  parentRefs:
+  - name: my-gateway
+  hostnames:
+  - foo.com
+  - bar.com
+  rules:
+  - matches:
+      method:
+        service: helloworld.Greeter
+        method:  SayHello
```

Comment:
method.method feels off, how about?
```suggestion
      service: helloworld.Greeter
      method:  SayHello
```

Reply:
Do note that there's another field here specifiying the _type_ of the match. Should the `method.type` field become a `methodType` field?

Reply:
I'd assumed this only supported an exact match, how would a prefix or regex match work when there are 2 values specified (service and method)?

Reply:
The only two supported types here are "exact" and "regular expression". Here's a full matrix of combinations and their equivalent URI matchers:

## Type Exact

|Service|Method|URI Matcher|
|----------|----------|-----------|
|Specified|Specified|Exact `/${SERVICE}/${METHOD}`|
|Specified|Unspecified|Prefix `/${SERVICE}/`|
|Unspecified|Specified|Suffix `/${METHOD}/` or Regex `/.+/${METHOD}`|
|Unspecified|Unspecified|Prefix `/`|

## Type RegularExpression

|Service|Method|URI Matcher|
|----------|----------|-----------|
|Specified|Specified|Regex `/${SERVICE}/${METHOD}`|
|Specified|Unspecified|Regex `/${SERVICE}/.+`|
|Unspecified|Specified|Regex `/.+/${METHOD}`|
|Unspecified|Unspecified|Prefix `/`|

I'll add this into the proposal to clear this up.

Reply:
Okay, that makes sense to me. Consider my objection resolved, but perhaps something addressing the stutter in the doc to preserve this for later?

Reply:
I agree with Nick and Harry here that the stutter does not feel quite right. The explanation here makes me mostly OK with it, but since I think this will be a common critique, it would be good to document the rationale in the GEP.